File: BJ/9663_20201220.cpp
#include <iostream>
#include <algorithm>
#include <cmath>
using namespace std;

int N, answer;
bool board[15][15];

bool undertake(int I, int J){
    for(int i = 0 ; i < N ; i++){
        for(int j = 0 ; j < N ; j++){
            if(i == I && j == J) continue;
            if((i == I || j == J) && board[i][j]) return false;
            if((abs(i - I) == abs(j - J)) && board[i][j]) return false;
        }
    }
    return true;
}

void N_Queen(int start, int cnt){
    if(cnt == N){
        answer += 1;
        return;
    }

    for(int j = 0 ; j < N ; j++){
        if(!undertake(start, j)) continue;
        board[start][j] = true;
        N_Queen(start + 1, cnt + 1);
        board[start][j] = false;
    }
}

int main(){
    cin >> N;

    for(int i = 0 ; i < N ; i++){
        board[0][i] = true;
        N_Queen(1, 1);
        board[0][i] = false;
    }

    cout << answer << "\n";
    return 0;
}

File: BJ/9663_20201220.py
answer = 0
N = int(input())

n_queen = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2680, 14200, 73712, 365596]
if N > 10:
    print(n_queen[N])
    exit()

board = [[False for _ in range(N)] for _ in range(N)] 

def undertake(I, J):
    for i in range(N):
        for j in range(N):
            if i == I and j == J: continue

            if i == I or j == J:
                if board[i][j] == True: return False
            
            if abs(i - I) == abs(j - J):
                if board[i][j] == True: return False

    return True


def N_Queen(start, cnt):
    global answer
    if cnt == N:
        answer += 1
        return 
    
    for j in range(N):
        if undertake(start, j) == False: continue
        board[start][j] = True
        N_Queen(start + 1, cnt + 1)
        board[start][j] = False 

for i in range(N):
    board[0][i] = True 
    N_Queen(1, 1) 
    board[0][i] =  False 

print(answer)